cmd/util/ledger/util/snapshot/payload_snapshot.go

```diff
@@ -266,3 +265,279 @@ func (p *indexMapSnapshot) ApplyChangesAndGetNewPayloads(
 
 	return result, nil
 }
+
+func workersToUse(workers int, payloads []*ledger.Payload) int {
+	// limit the number of workers to prevent too many small jobs
+	const minPayloadsPerWorker = 10000
+	if len(payloads)/minPayloadsPerWorker < workers {
+		workers = len(payloads) / minPayloadsPerWorker
+	}
+	if workers < 1 {
+		workers = 1
+	}
+	return workers
+}
+
+func createLargeIndexMap(
+	log zerolog.Logger,
+	payloads []*ledger.Payload,
+	workers int,
+) (map[flow.RegisterID]int, error) {
+
+	workers = workersToUse(workers, payloads)
+
+	if workers == 1 {
+		r := make(map[flow.RegisterID]int, len(payloads))
+		for i, payload := range payloads {
+			key, err := payload.Key()
+			if err != nil {
+				return nil, err
+			}
+			id, err := convert.LedgerKeyToRegisterID(key)
+			if err != nil {
+				return nil, err
+			}
+			r[id] = i
+		}
+
+		return r, nil
+	}
+
+	start := time.Now()
+	log.Info().
+		Int("payloads", len(payloads)).
+		Int("workers", workers).
+		Msgf("Creating payload snapshot")
+
+	defer func() {
+		log.Info().
+			Int("payloads", len(payloads)).
+			Int("workers", workers).
+			Dur("duration", time.Since(start)).
+			Msgf("Created payload snapshot")
+	}()
+
+	type result struct {
+		minimap map[flow.RegisterID]int
+		err     error
+	}
+
+	minimaps := make(chan result, workers)
+	defer close(minimaps)
+
+	// create the minimaps in parallel
+	for i := 0; i < workers; i++ {
+		start := i * len(payloads) / workers
+		end := (i + 1) * len(payloads) / workers
+		if end > len(payloads) {
+			end = len(payloads)
+		}
+
+		go func(startIndex int, payloads []*ledger.Payload) {
+			minimap := make(map[flow.RegisterID]int, len(payloads))
+			for i, payload := range payloads {
+				key, err := payload.Key()
+				if err != nil {
+					minimaps <- result{nil, err}
+					return
+				}
+				id, err := convert.LedgerKeyToRegisterID(key)
+				if err != nil {
+					minimaps <- result{nil, err}
+					return
+				}
+				minimap[id] = startIndex + i
+			}
+			minimaps <- result{minimap, nil}
+		}(start, payloads[start:end])
+
+	}
+
```

Comment:
Maybe have a shortcut here if `workers == 2` to avoid goroutine overhead of paring and merging mini maps.

Same comment for `createLargeMap()`.
 
```Go
if workers == 2 {
    left := <-minimaps
    if left.err != nil {
        return nil, left.err
    }

    right := <-minimaps
    if right.err != nil {
        return nil, right.err
    }

    maps.Copy(left.minimap, right.minimap)
    return left.minimap, nil
}
```

Reply:
Looks like this was addressed in [aea4edc](https://github.com/onflow/flow-go/pull/5646/commits/aea4edc3efd07e6d61bf6e319a8cce1595c775ac)